timeseries.py

```diff
@@ -0,0 +1,650 @@
+        conf_lo = self._combine_or_none(self._confidence_lo, other.conf_lo_pd_series(), combine_fn)
+        conf_hi = self._combine_or_none(self._confidence_hi, other.conf_hi_pd_series(), combine_fn)
+        return TimeSeries(series, conf_lo, conf_hi)
+
+    """
+    Definition of some methods related to seasonality.
+    """
+
+    def check_seasonality(self, m: int = 0) -> Tuple[bool, int]:
+        """
+        Checks if the TimeSeries has a seasonality of order [m].
+
+        :param m: An integer, the order of seasonality to check.
+        :return: A tuple (a, b), where a indicates whether there is seasonality of order [m] or not and b = m
+                unless the TimeSeries is constant.
+        """
+        n_unique = np.unique(self.values()).shape[0]
+        if n_unique > 1:
+            r = acf(self.values(), nlags=max([m, 24]))  # In case user-defined period exceeds 24
+            if m == 0:  # Handles the case with unknown seasonality period
+                grad = np.gradient(r)
+                signs_changes = np.diff(np.sign(grad))
+                m = np.nonzero((signs_changes == -2))[0][0] + 1  # +1 correction as np.nonzero returns an index.
+            r = r[1:]  # Removes the auto-correlation with lag 0.
+            stat = self.compute_stats(r, m, len(self))
+            return (abs(r[m-1]) / stat) > norm.ppf(0.95), m
+        return False, 0
+
+    @staticmethod
+    def compute_stats(r: np.ndarray, m: int, N: int) -> float:
+        """
+        Computes the standard error of [r] with order [m] for a sample of size [N] according to Bartlett's formula.
+
+        :param r: A numpy array.
+        :param m: An integer, the upper limit on the number of terms of r to use.
+        :param length: An integer, the length of the sample to test.
+        :return: A float, the standard error of [r] with respect to [m] and [length].
+        """
+        if m == 1:
+            return math.sqrt(1/float(N))
+        else:
+            return math.sqrt((1 + 2 * sum(map(lambda x: x ** 2, r[:m - 1]))) / float(N))
+
+    def seasonal_adjustment(self, model: str = 'multiplicative') -> Tuple['TimeSeries', 'TimeSeries']:
+        """
+        Returns the TimeSeries, adjusted for seasonality and the seasonality TimeSeries.
+
+        :param model: A string, the type of seasonality to consider (additive or multiplicative).
+        :return: A Tuple (a,b) of two TimeSeries, where a is obtained from this TimeSeries after seasonal adjustment
+                and b is the seasonality TimeSeries itself.
+        """
+        decomp = seasonal_decompose(self.pd_series(), model=model)
+        seasonality = decomp.seasonal
+        if (seasonality < 1e-6).any():
+            print("WARNING seasonal indexes equal to zero, using non-seasonal Theta method.")
+            ts = self
+        else:
+            if model == 'multiplicative':
+                ts = TimeSeries.from_times_and_values(self.time_index(), self.values() / seasonality)
```

Comment:
how about the case where we have confidence intervals?
How about using the dunder methods that you defined below?